service/history/chasm_engine.go

```diff
@@ -241,14 +285,182 @@ func (e *ChasmEngine) ReadComponent(
+
+	shardContext, executionLease, err := e.getExecutionLease(ctx, requestRef)
+	if err != nil {
+		// E.g. requestRef VT inconsistent with shard VT ('stale reference')
+		return nil, err
+	}
+
+	// hold lock until return or subscription established
+	released := false
+	defer func() {
+		if !released {
+			// read-only operation; don't pass error
+			executionLease.GetReleaseFn()(nil)
+		}
+	}()
+
+	// At this point it's possible that shard VT < requestRef VT (getExecutionLease does not
+	// guarantee that returned shard state is non-stale w.r.t. requestRef).
+	currentRef, err := e.checkPredicate(ctx, requestRef, executionLease, predicateFn)
+	if err != nil {
+		return nil, err
+	}
+	// shard VT >= requestRef VT (enforced by checkPredicate)
+
+	if currentRef != nil {
+		// wait condition was satisfied
+		return currentRef, nil
+	}
+
+	// Wait condition not satisfied; long-poll
+
+	// TODO(dan) For now, UpdateComponent emits execution-level notifications, and PollComponent
+	// subscribes to execution-level notifications. This means that PollComponent may be woken up
+	// unnecessarily, but it will not miss notifications. In the future we may want to change
+	// PollComponent to subscribe to component-level notifications, and change UpdateComponent so
+	// that notifications are emitted only for nodes that were mutated in the transaction.
+	channel, subscriberID, err := e.notifier.Subscribe(requestRef.EntityKey)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = e.notifier.Unsubscribe(requestRef.EntityKey, subscriberID)
+	}()
+
+	// Release the lock, now that we are subscribed
+	executionLease.GetReleaseFn()(nil)
+	released = true
+
+	namespaceRegistry, err := shardContext.GetNamespaceRegistry().GetNamespaceByID(
+		namespace.ID(requestRef.EntityKey.NamespaceID),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Long-poll timeout semantics are the same as PollWorkflowExecutionUpdate: if the long-poll
+	// times out due to the server-imposed soft timeout, then return a non-error empty response
+	// indicating to a caller that they should continue long-polling. Unlike
+	// PollWorkflowExecutionUpdate we reserve 1s buffer.
+	softTimeout := shardContext.GetConfig().LongPollExpirationInterval(namespaceRegistry.Name().String())
+	stCtx, stCancel := contextutil.WithDeadlineBuffer(ctx, softTimeout, time.Second)
+	defer stCancel()
+
+	for {
+		select {
+		case notification := <-channel:
+			if notification == nil {
+				return nil, serviceerror.NewInternal("nil component notification")
+			}
+			// Received a notification. Re-acquire the lock and check the predicate.
+			_, executionLease, err := e.getExecutionLease(stCtx, requestRef)
+			if err != nil {
+				if ctx.Err() != nil && errors.Is(err, ctx.Err()) {
+					return nil, ctx.Err()
+				}
+				if errors.Is(err, stCtx.Err()) {
+					// Server-imposed timeout; caller should continue long-polling.
+					return nil, nil
+				}
+				return nil, err
+			}
```

Comment:
It's common to ignore the error returned check `ctx.Err() != nil` instead. This allows the implementor of the called function to propagate any error it wishes and doesn't bind it to returning the specific context deadline error as the contract.

I also don't know what you'd gain from checking the original context's error, at that point, it's likely that the frontend will have timed out the request and whatever you return would be ignored.

The following code is much simpler to follow:

```suggestion
			if errors.Is(err, stCtx.Err()) {
				// Server-imposed timeout; caller should continue long-polling.
				return nil, nil
			}
			return nil, err
```

You can save some complexity by assigning the new context to the same `ctx` variable to remove the need to reason about which context should be referenced where.